Read-Only JSON API Tutorial

izdwuut opened the pull request:
This is a 🔦 documentation change. 

## Summary

I've added a tutorial on building Read-Only JSON APIs. It mostly covers creating a custom plugin with generators, converters and hooks. This is a detailed step-by-step guide. 

I think it's great that you allow such contributions. I was about to put it behind a paywall on Medium, but if you add the tutorial to the docs, it could reach far more people.

[Preview](https://deploy-preview-8521--jekyllrb.netlify.app/tutorials/read-only-json-api/) 👀

DirtyF commented:
Right, ideally our docs should have four separated parts:
- tutorials (get a project done)
- how-tos (get a task done)
- reference (options, tags, filters,) with examples
- explanations (high-level  understanding of some concepts)

izdwuut commented:
@DirtyF, I believe that my contribution would fall under the first category you've listed. Is there something I've got wrong?

DirtyF commented:
@izdwuut Not at all, it's mostly a reminder for me to reevaluate how we should organize Jekyll documentation.

MichaelCurrin commented:
Thanks for the contribution.

Here is the deploy preview link for ease of reading.

https://deploy-preview-8521--jekyllrb.netlify.app/tutorials/read-only-json-api/

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,321 @@
+date: 2020-12-23 10:04:00 +0100
+---
+
+There shouldn't be a necessity to explain what is the [API](https://en.wikipedia.org/wiki/API). I assume that you've landed here to learn it in a context of Jekyll. Let's jump straight to the actual thing, shall we?
```
I think the tone and the info here can be improved to not presume to much and also to tell the reader in the first line what the article is going to create and why they should care.

e.g.

```
Here we take your posts markdown files in your Jekyll site and make them available as static JSON files on a kind of REST API. Then you can view them directly in the browser as `/foo.json` and you can use them in your frontend layer, such as if you have a JavaScript application on the frontend which pulls in JSON content.

We are using [JSON](https://en.wikipedia.org/wiki/JSON) is a popular cross-language method of transferring data on the web, without all the presentation and styling of HTML. The files we output will be available in a similar way to if you built a [REST](https://restfulapi.net/) [API](https://en.wikipedia.org/wiki/API) in a server-side language like Ruby, Python or Node. 
```

izdwuut replied:
Fixed. yeah, I agree that tone is unpleasant. It just that I've been too blunt about caring about the reader time.

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,321 @@
+First, you need to go through an [installation]({% link _docs/installation.md %}) process. Having that, you can initialize a new Jekyll project. In your directory of choice, invoke the following:
+
+```sh
+bundle init
````
Can you find a tutorial on the site that covers these steps already as a central source of truth? 

It is not nice to duplicate similar but slightly different setup steps across tutorials, as they don't follow a convention and they all have to be updated if a change is needed.

Instead of covering setup steps and things that the user _won't_ need in the Config section, could you just outline what the user _does_ need? 

e.g.

```
_layouts/
  post.html
_posts/
  2020-12-12-hello.html
index.md
_config.yml
```

Then they can use the jekyll new command if they want or create the files by hand.

izdwuut replied:
Fixed. You're right. It's better to define something once and reference it whenever possible.

MichaelCurrin commented:
I wonder if the solution here is overly complicated. Especially at it relies on Ruby code as custom plugins, which is a barrier to setup and maintain. And can't be used on GH Pages due to restricted use of plugins.

For comparsion, I've made tutorial gist which uses builtin Jekyll functionality to make JSON files from a page and a post, without using any Ruby code. It might to everything covered in your case, but I think it much easier to understand and implement for the average Jekyll user.

https://gist.github.com/MichaelCurrin/f8d908596276bdbb2044f04c352cb7c7

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,321 @@
+
+---
+
+# Configuration
```
Rather increase `#` to `##`. And existing `##` to `###`.

Having multiple H1 headings on an HTML page is not a good pattern for styling, accessibility and probably SEO. It is also not consistent with other pages like [navigation](https://raw.githubusercontent.com/jekyll/jekyll/1d4cff612bcfe28b5dc1be8331b6a47bc3b909f1/docs/_tutorials/navigation.md) tutorial which uses H2s, no H1 and doesn't use `---` to split sections.

izdwuut replied:
Done. Wasn't aware of that.

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,321 @@
+Before you do it, it's worth noting that you can list all the available properties with this little template, which returns an array of available keys:
+
+```liquid
+{% raw %}{{ page.keys | jsonify }}{% endraw %}
````
I don't know how other maintainers feel, but I find the code much easier to read and maintain by wrapping the entire file in `raw`.

```
---
---
{% raw %}

## Heading

content

...

Bottom of page

{% endraw %}
```

izdwuut replied:
Fixed. I agree, it's nicer, although I get a warning when I do it:

```
Excerpt modified in _tutorials/read-only-json-api.md!
Found a Liquid block containing the excerpt separator "\n\n". 
The block has been modified with the appropriate closing tag.
Feel free to define a custom excerpt or excerpt_separator in the
document's Front Matter if the generated excerpt is unsatisfactory.
```
Any idea how to resolve it? It's just a warning, but I feel that it should be addressed anyways.

MichaelCurrin replied:
Hmm I have used raw tag frequently including for an entire page and have not seen that warning before.

An excerpt is usually in the context of a post but I guess applies to this tutorial too. The excerpt is the first paragraph and it will have a blank line after it which is the \n\n

After rewriting the opening paragraph if you still get this warning you can try play around with whitespace. 

E.g.


```
---
---

{% raw %}
There is...
```

Or 

```
---
---
{% raw %}There
```
Or maybe just start the raw tag after the first paragraph.


```
---
---
First paragraph here

{% raw %}
Second paragraph and rest of tutorial here
...

{% endraw %}
```

izdwuut replied:
Starting the raw tag after the first paragraph fixed the warning. But, I noticed that because the whole post is wrapped in the tag, URLs to internal pages are not generated properly.

MichaelCurrin replied:
Sorry I should have checked on what would break before making the suggestion 

You'll have to use the raw tag throughout but I would recommend keeping it readable.

e.g. not squeezed in one line 

```
{% raw %}Text
```

And not this either before raw tag actually renders as an empty line

    ```liquid
    {% raw %}
    Text
    ```


But this should be good with raw on the outside.

    {% raw %}
    ```liquid
    Text

izdwuut replied:
Done.

izdwuut commented:
> I wonder if the solution here is overly complicated. Especially at it relies on Ruby code as custom plugins, which is a barrier to setup and maintain. And can't be used on GH Pages due to restricted use of plugins.
Yes, the fact that it can't be hosted as-is on GH Pages is a major downside of my approach. Theoretically you could generate your page yourself and put it on `gh-pages` branch (and keep your sources on another branch), but it adds yet another layer of complexity to already complex solution.

> For comparsion, I've made tutorial gist which uses builtin Jekyll functionality to make JSON files from a page and a post, without using any Ruby code. It might to everything covered in your case, but I think it much easier to understand and implement for the average Jekyll user.
> 
> https://gist.github.com/MichaelCurrin/f8d908596276bdbb2044f04c352cb7c7

This is interesting. It didn't strike me at first that the problem can be tackled using layouts alone. The only thing I don't like about your approach is that you rely on many JSON files that, in reality, are either Markdown or HTML files - this is especially valid for posts. In VSCode - according to my knowledge - it would result in not being able to dynamically preview your post. Since it has a `.json` extension, it just wouldn't be recognized as a valid Markdown document - if I'm not wrong. In my solution, there is only one file that behaves like this - a `default.html` layout.

Would you mind if I linked to your Gist in my tutorial? To some, it could be a nice (and probably preferred) alternative.

MichaelCurrin commented:
> > I wonder if 
> Yes, the fact that it can't be hosted as-is on GH Pages is a major downside of my approach. Theoretically you could generate your page yourself and put it on `gh-pages` branch (and keep your sources on another branch), but it adds yet another layer of complexity to already complex solution.

What you can do make a mention that the solution needs a non-standard environment to support the custom ruby code and then link to the GH Actions deploy tutorial. It's not under tutorials but you can figure out the path and use the `link` tag.

https://jekyllrb.com/docs/continuous-integration/github-actions/

MichaelCurrin commented:
@izdwuut you can configure VS Code file associations.

You can do it temporarily for any given file (which will reset when you close an open the file). Click the language in the bottom right e.g. JSON and choose "Jekyll (HTML)" for liquid highlighting or Markdown (usually for .md files).
You can do it more permanently using Configure File Associations or similar in the command prompt menu.

You can also configure this in your project's `.vscode/settings.json` file so it applies to all JSON files but doesn't affect your other projects.

e.g.
```
{
    "files.associations": {
        "*.json": "jekyll"
    }
}
```

And yes please do. I'll keep the gist around so you are welcome to link to it. Thanks.

izdwuut commented:
I've introduced all the changes. In one of the commits I've added a disclaimer where I inform about GitHub Actions and link to your Gist.

ashmaroli commented:
@izdwuut I'd recommend adding one more *disclaimer* at the very end of the tutorial something along the lines of:
```
Disclaimer: The Ruby code in this tutorial is only for illustration purpose. The Jekyll team is not be liable to resolve
any bugs or issues arising from its use in production.
```

izdwuut commented:
@ashmaroli I believe it belongs to the generic [tutorials](https://jekyllrb.com/tutorials/home/) page. This could apply to any tutorial added in the future, so I think it's better to to put it there. What do you think?

ashmaroli commented:
> I believe it belongs to the generic [tutorials](https://jekyllrb.com/tutorials/home/) page. This could apply to any tutorial added in the future, so I think it's better to to put it there.

@izdwuut Since the generic tutorial page has a lower chance of getting visitors compared to individual tutorials, I believe it'd be better to have the disclaimer rendered on necessary tutorials for greater visibility.

See proposal at #8525

izdwuut commented:
This is a nice, clean solution. Since it's very likely to get accepted, I'll just add the field you propose to the front matter in my tutorial.

chrisfinazzo commented:
👋🏼 @izdwuut, good to have you.

I'll give this a look later today, mostly for copyediting purposes.

ashmaroli commented:
Glad to see you after long time, @chrisfinazzo :smiley: 
Thank you!!

izdwuut commented:
@chrisfinazzo Thanks!

Oh yes, that would be nice, since English is not my native language. I'm so glad that you can offer such help!

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,307 @@
+author: izdwuut
+date: 2020-12-23 10:04:00 +0100
+---
+In this tutorial, you'll create a kind of a [REST](https://restfulapi.net/) [API](https://en.wikipedia.org/wiki/API) serving static [JSON](https://www.json.org/json-en.html) files generated from Markdown posts in your Jekyll site.
```
@izdwuut Much better, thanks for updating :)

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,307 @@
+Let's dive in!
+
+{: .note .info}
+This solution needs a custom environment to support the Ruby code. This means that you can't host the site on GitHub Pages as-is - you'd need to use [Actions]({% link _docs/continuous-integration/github-actions.md %}) to deploy your site. First, see if a simpler [solution](https://gist.github.com/MichaelCurrin/f8d908596276bdbb2044f04c352cb7c7) created by [Michael Currin](https://github.com/Michae ...
```
:+1:

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,307 @@
+You'll get this in `/_site/update/2020/12/20/the-first-post.html`:
+
+```json
+["url","relative_path","excerpt","title","categories","path","next","previous","id","output","date","tags","content","collection","draft","layout","category","custom-property","slug","ext"]
````
Some wrapping for readability.

```suggestion
["url","relative_path","excerpt","title","categories","path","next",
"previous","id","output","date","tags","content","collection",
"draft","layout","category","custom-property","slug","ext"]
```

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,307 @@
+After building the site, you'd get the following:
+
+```json
+{"title":"The First Post","categories":["update"],"tags":[],"content":"<p>1</p>\n","collection":"posts","date":"2020-12-20 00:00:00 +0100","custom-property":"My custom property"}
````
```suggestion
{"title":"The First Post","categories":["update"],"tags":[],
 "content":"<p>1</p>\n","collection":"posts",
 "date":"2020-12-20 00:00:00 +0100","custom-property":"My custom property"}
```

MichaelCurrin reviewed (changes requested):
I made some style suggestions but otherwise looks fine.

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,307 @@
+
+    category_dir = site.config['category_dir'] || 'categories'
+    site.categories.each_key do |category|
+      site.pages << ListingPage.new(site, site.source, File.join(category_dir, category), categories[category])
```
The code cuts off around `File.join`.

https://deploy-preview-8521--jekyllrb.netlify.app/tutorials/read-only-json-api/#generate-pages

```suggestion
      site.pages << ListingPage.new(site, site.source,
                                    File.join(category_dir, category), 
                                    categories[category])
```

Or

```ruby
foo.new(
  bar
  baz
  buzz
)
```

MichaelCurrin commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,307 @@
+      site.pages << ListingPage.new(site, site.source, File.join(category_dir, category), categories[category])
+    end
+    site.pages << ListingPage.new(site, site.source, "", posts)
+    site.pages << ListingPage.new(site, site.source, category_dir, categories.keys)
```
```suggestion
    site.pages << ListingPage.new(site, site.source, category_dir, 
                                  categories.keys)
```

izdwuut commented:
@MichaelCurrin Nice to hear it. I've introduced the changes you've proposed.

izdwuut commented:
Since #8525 is now merged, I've added the appropriate field to a front matter to display the disclaimer.

ashmaroli commented:
@izdwuut We have also updated the [documentation on generators](https://jekyllrb.com/docs/plugins/generators/)..

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+serving static [JSON](https://www.json.org/json-en.html) files generated from Markdown posts in your Jekyll site.
+
+JSON is a popular cross-language method of transferring data on the web, without all the presentation layer of HTML.
+The files you output will be similar to a REST API in a server-side language like Ruby, Python or Node. You'll be able
```
"The files you output will be similar to a REST API in a server side language like Ruby, Python, or Node."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+JSON is a popular cross-language method of transferring data on the web, without all the presentation layer of HTML.
+The files you output will be similar to a REST API in a server-side language like Ruby, Python or Node. You'll be able
+to see them directly in the browser as `/foo.json`. They can be used to build a custom frontend using Angular, React,
```
Same as above.

"Angular, React, Vue, and so on"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+by [Régis Philibert](https://forestry.io/authors/r%C3%A9gis-philibert/). It's about doing the exact same thing in
+[Hugo](https://gohugo.io/).
+
+I hope that this solution in Jekyll is elegant, too. You just need to define some [layouts]({% link _docs/layouts.md %})
```
"You will need to define a number of layouts and write a plugin to convert your content to JSON."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+I hope that this solution in Jekyll is elegant, too. You just need to define some [layouts]({% link _docs/layouts.md %})
+and write a [plugin]({% link _docs/plugins.md %}) to convert your content to JSON. **Don't be afraid if you're not
+familiar with Ruby** --- the language that Jekyll is written in. It's similar to Python, and you'll get a hang of it
```
"...you should get the hang of it quickly"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+Let's dive in!
+
+{: .note .info}
```
"...to support its Ruby code."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+## Setup
+
+First, you need to go through an [installation]({% link _docs/installation.md %}) process. Having that, you can
```
"Having done that," etc. etc.

chrisfinazzo replied:
"invoking the command `bundle exec jekyll new`."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,354 @@
+```
+
+Unless you self-host your site, **you can rely on a generic 404 response from your server**. You no longer need `index`
+ --- you will generate a custom one later. Generating pages like `about` is out of scope of this tutorial, but in case
````
"...out of scope for this tutorial"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+ --- you will generate a custom one later. Generating pages like `about` is out of scope of this tutorial, but in case
+you need them, they behave the same way that posts do.
+
+Under `_posts`, create two new documents. Remember that Jekyll requires you to **prepend post filename with a date**:
```
"post filenames"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,354 @@
+1
+```
+
+Please note that you include a `custom-property` that you will fetch later on. 
````
"Make sure to include a `custom-property` so you can fetch it later on"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,354 @@
+2
+```
+
+This one belongs to two categories --- `update` and `tutorial`, and doesn't include the custom property. Because the
````
"This one belongs to two categories  - `update` and `tutorial` - and doesn't include the custom property.

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+[cover]({% link _docs/configuration/front-matter-defaults.md %}) every file.
+
+## Convert Posts to JSON
+My approach is a little hacky --- it requires you to specify a JSON output format in a HTML template. The only thing
```
"...in an HTML template."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+## Convert Posts to JSON
+My approach is a little hacky --- it requires you to specify a JSON output format in a HTML template. The only thing
+you miss is that your text editor will most probably not recognize the markup file as a properly formatted JSON
```
"The only thing you will miss is that your text editor will probably not recognize the markup file as a properly formatted JSON document"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+{% endraw %}
+
+* Pass every property through a jsonify [filter]({% link _docs/liquid/filters.md %}#data-to-json). This will add
+quotation marks for us, convert arrays into strings and escape quotation marks.
```
"This will add quotation marks for us, convert arrays into strings, and escape quotation marks."

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+* Pass every property through a jsonify [filter]({% link _docs/liquid/filters.md %}#data-to-json). This will add
+quotation marks for us, convert arrays into strings and escape quotation marks.
+* For content, turn Markdown-formatted string into HTML using [markdownify]({% link _docs/liquid/filters.md %}#markdownify)
```
"Markdown-formatted strings"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+quotation marks for us, convert arrays into strings and escape quotation marks.
+* For content, turn Markdown-formatted string into HTML using [markdownify]({% link _docs/liquid/filters.md %}#markdownify)
+filter.
+* Fetch the falue of custom-property variable that you've defined earlier. If it doesn't exist on a post, a null value
```
"...the custom variable you defined earlier"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+* Fetch the falue of custom-property variable that you've defined earlier. If it doesn't exist on a post, a null value
+will be returned.
+
+If you enter `bundle exec jekyll build` command again, you'd get the following output:
```
"If you run `bundle exec jekyll build` again, you'll get the following  output:"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+### Create the Plugin
+
+The rest of the work can be done using a custom plugin. For now, it only needs to convert the output file extension
```
"filename extension"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+* `convert` --- a function that takes post content **transpiled into HTML** as an argument. Please note that it
+doesn't parse the file through a template.
+
+### Remove Whitespaces
```
Remove Whitespace

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+### Remove Whitespaces
+
+But, there's another trick. Using [hooks]({% link _docs/plugins/hooks.md %}), you can remove all whitespaces! Before a
```
"...all whitespace!"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+### Remove Whitespaces
+
+But, there's another trick. Using [hooks]({% link _docs/plugins/hooks.md %}), you can remove all whitespaces! Before a
+post is rendered, you can squeeze some custom logic in:
```
"squeeze" is awkward here, perhaps you you say "add" or something similar?

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+
+### Define a Page
+
+In `api_generator.rb`, add the following class as a part of `Jekyll` module:
```
"...the `Jekyll` module"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+* Assign some [instance variables](https://www.ruby-lang.org/en/documentation/faq/8/) that Jekyll will utilize to
+generate the page.
+* Process the page filename. It will extract the extension and base name.
+* `read_yaml` is only needed to define some instance variable for us. It's just simpler than doing it by hand. You can
```
"instance variables"

"It's simpler than doing it by hand"

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,354 @@
+* Insert the post to `categories` and `posts` hashes.
+* Generate categories, categories index and the index of all posts (in that order)
+
+And that's it! If you hit `bundle exec jekyll build` again, you'd get the following output:
```
"you'll get the following output:"

chrisfinazzo reviewed (changes requested):
I think this is almost ready - see comments - mostly phrasing and readability tweaks, but otherwise it's good.

In terms of tone, I was tempted to remove the first-person ("I") and swap in the "royal we" (as in The Jekyll Project), but realized that these are @izdwuut's words, not mine or anyone else's, so that's how it should be.

Don't feel bad about not getting this exactly right. After all, [English is a stupid language][].

[English is a stupid language]: https://www.youtube.com/watch?v=Q8mD2hsxrhQ

izdwuut commented:
@chrisfinazzo Oh yes, English language can get tricky.  I'm glad that your suggestions were only minor tweaks. It feels reassuring :).

About writing it in first person: yeah, I wasn't sure either. I realized that this can be fixed later and didn't put too much thought to it.

@ashmaroli I have troubles including the changes introduced in generators page, but I'm working on it. I still can't figure how to enable custom permalinks and generate list of categories. I have an idea, but it would require me to writing some boilerplate code, which isn't nice. I'll keep you informed.

ashmaroli commented:
@izdwuut The reason I pointed you towards the updated doc on generators is to highlight the fact that calling the `read_yaml` method is not a requisite for pages *generated at runtime*. Since there are no physical files, there doesn't arise a need to *read* something. Also, reading the same physical layout file for every generated page is wasteful IO operation even on the fastest computer.
The core `Jekyll` module also has the `PageWithoutAFile` class that omits the IO aspect.
For additional reference, you may look into existing plugins such as `jekyll-archives`, `jekyll-feed`, `jekyll-sitemap`, etc.

izdwuut commented:
@ashmaroli I'll definitely take a look at this. But, when it comes to the docs, the new generator page is nice. It provides a way to externalize config and gets rid of `read_yaml`, too. I think I like that approach more. I eventually figured out how to apply the changes to my solution, so it would be no hassle to update the tutorial. It's just that it would be necessary to rewrite a major part of it.

Just getting rid of `read_yaml` would be the simplest way, but I believe that some core changes would be more benefitial. I'm okay with both approaches, so the decision is ultimately yours.

ashmaroli commented:
@izdwuut  You're free to choose the route that is most acceptable to you, as the author, given the information at your disposal.
The maintainers would only arbitrate the language and reading-flow aspect of the tutorial.

izdwuut commented:
@ashmaroli I chose to include changes introduced in generators page. It makes the solution more extensible.

ashmaroli commented:
@izdwuut Great to know that, but why is there a `post.rb` inside the `_includes` directory?

izdwuut commented:
@ashmaroli Morning grog. I edited the tutorial so it relies on more templates (that include the partial), but I forgot to describe it. It should be okay now.

DirtyF commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,421 @@
+In this tutorial, you'll create a kind of a [REST](https://restfulapi.net/) [API](https://en.wikipedia.org/wiki/API)
+serving static [JSON](https://www.json.org/json-en.html) files generated from Markdown posts in your Jekyll site.
+
+JSON is a popular cross-language method of transferring data on the web, without all the presentation layer of HTML.
```
```suggestion
JSON is a popular format to structure and exchange data on the web, without all the presentation layer of HTML.
```

DirtyF commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,421 @@
+serving static [JSON](https://www.json.org/json-en.html) files generated from Markdown posts in your Jekyll site.
+
+JSON is a popular cross-language method of transferring data on the web, without all the presentation layer of HTML.
+The files you output will be similar to a REST API in a server side language like Ruby, Python, or Node. You'll be able
+to see them directly in the browser as `/foo.json`. They can be used to build a custom frontend using Angular, React, 
```
```suggestion
The files you output are similar to a REST API in a server side language like Ruby, Python, or Node. You
can access them directly in your browser from e.g. `/foo.json`. They can be used to build a custom frontend using Angular, React, 
```

DirtyF commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,421 @@
+Vue, and so on.
+
+A huge inspiration to write this tutorial was a [post](https://forestry.io/blog/build-a-json-api-with-hugo/) written
+by [Régis Philibert](https://forestry.io/authors/r%C3%A9gis-philibert/). It's about doing the exact same thing in
```
```suggestion
by [Régis Philibert](https://twitter.com/regisphilibert). It's about doing the exact same thing in
```

DirtyF commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,421 @@
+Let's dive in!
+
+{: .note .info}
+This solution needs a custom environment to support it's Ruby code. This means that you can't host the site on GitHub
```
```suggestion
This solution needs a custom environment to support its Ruby code. This means that you can't host the site on GitHub
```

DirtyF commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,421 @@
+
+### Directories Structure
+
+You need to create the following structure:
```
A public repo would be nice here.

DirtyF reviewed (approved):
A few minor comments, thanks for this tutorial @izdwuut, let us know if we can point to a public repository that would make it easier to follow this through.

ashmaroli commented:
> I edited the tutorial so it relies on more templates

@izdwuut I still don't see `_includes/post.rb` being used anywhere..

izdwuut commented:
@ashmaroli Right! I see what you mean now. This file should actually be `post.html`, as it's just a HTML template. Sorry for the confusion. At first I thought that the filename is right and I've just forgot to describe it.

@DirtyF I'm going to add it in a bit. I have changes to push and I'm not in front of my desktop PC right now.

DirtyF commented:
@izdwuut No hurry, we'll publish when it's done 😄

izdwuut commented:
I updated the description for `ApiGenerator` class to utilize a Ruby `map` instead of a regular loop. There also lied a bug, which I fixed.

Also, I [created](https://github.com/izdwuut/jekyll-json-api-tutorial) the repo for the project. Any idea where to include it to have the most impact? Should it be a note or is it enough to link it a place you've pointed in your review, @DirtyF?

ashmaroli commented:
@izdwuut You seem to have introduced typos to the document in [the latest commit](https://github.com/jekyll/jekyll/pull/8521/commits/b085540e1e06ce381c85534fdbf5eb1c8322b757).
Regarding the repo, you can replace the directory structure in this tutorial with a link to the repo. Also, it can be whatever licence you wish. You are the author and owner of the repo.

DirtyF commented:
@izdwuut I would mention the repository link from the beginning (setup for people to be able to follow through or for the impatient that was to see the code directly.

Maybe you can invite people to clone the repository first and adapt the text copy?

izdwuut commented:
@DirtyF Yes, I think I can add a branch with barebone directories structure and ask users to clone it first.

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,416 @@
+2
+```
+
+This one belongs to two categories --- `update` and `tutorial` --- and doesn't include the custom property. Because the
````
```suggestion
This post belongs to two categories --- `update` and `tutorial` --- and doesn't include the custom property. Because the
```

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,416 @@
+```
+
+This one belongs to two categories --- `update` and `tutorial` --- and doesn't include the custom property. Because the
+categories overlap, it will be possible to test getting a list of posts from a category.
````
```suggestion
categories overlap with that of the previous post, it will be possible to test getting a list of posts from a category.
```

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,416 @@
+
+### Layouts
+
+The last thing is to define the layouts. First, declare them in `config.yml`:
```
Too early to talk about doing *the last thing*. We haven't created the Ruby plugin yet.
Besides, using front matter defaults and new types will only be clear after the Ruby plugin has been implemented.

So it's best to move this section to the bottom.

izdwuut replied:
Done. I also made some changes to this section, so that it reflects the fact that it's the last thing I describe.

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,416 @@
+      layout: items_index
+```
+
+* You have four various content types: an index of all categories, a listing of all posts under a category, a 
````
Four various content types?
The `type` key doesn't denote *page content type*.

Since an index is a listing as well, using both terms here can be confusing. Either there are three listings or three indexes.

izdwuut replied:
I updated this line. Now I reference it as `scopes`, which I believe is more accurate.

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,416 @@
+ "draft","layout","category","custom-property","slug","ext"]
+```
+
+Given that, I'd like you to enter the following template:
````
The use of the term *enter* in this line and the previous ones is incorrect.
Conventionally, one is instructed *to enter a location* or *enter data into a document*.
*Entering a template* can therefore be misinterpreted.

Additionally, it is not clear where to insert the subsequent Liquid code..

ashmaroli commented on docs/_tutorials/read-only-json-api.md:
````diff
@@ -0,0 +1,416 @@
+{% raw %}
+```liquid
+{
+    "title": {{ include.post.title | jsonify }},
````
what happens if one were to use `"title": {{ page.title | jsonify }}` instead?
`

izdwuut replied:
A listing of posts under a category would render with `null` values.

ashmaroli commented:
@izdwuut I understand that English is not your first language. So, I'm converting this to a draft PR.
Feel free to take your time, finish the entire document, read through the entire tutorial by *serving the docs site locally* and then push the changes.

Ping us when you are ready.

chrisfinazzo commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,416 @@
+familiar with Ruby** --- the language that Jekyll is written in. It's similar to Python, and you should get a hang of it
+quickly.
+
+By the end of the tutorial, you'll expose the following API:
```
"endpoints", no?

izdwuut commented:
I'm terribly sorry for forcing you to review the PR in the previous state. I have to admit it - it was awful at places, so I'm thankful for your patience. I updated the tutorial according to your remarks. I read the tutorial a couple of times and made sure that people who follow it can actually yield the result. There still may be errors, but I did my best to get rid of most of them.

I hope it's okay if I ping just you, @ashmaroli?

izdwuut commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,416 @@
+Let's dive in!
+
+{: .note .info}
+This solution needs a custom environment to support its Ruby code. This means that you can't host the site on GitHub
```
I updated this paragraph. I think it's less confusing now.

ashmaroli commented:
@izdwuut I'm sorry to force more edits. But recent commits seem to have removed chunks of text.
Please read through the section titled `### Dummy Posts`...

izdwuut commented:
@ashmaroli It's okay. It's your job, after all. 

It's probably because I've had conflicts when merging and resolved the changes incorrectly. You're right - there are places where things are not right. I'll do my best to fix it in a reasonable time. It's not that we're in a hurry, but this discussion is exceptionally long. You probably have better things to do (and more important, too) than taking care of this tutorial for another week.

izdwuut commented:
@ashmaroli I know why this problem arose. It's because there were some changes that you - the maintainers team - proposed that I introduced _after_ fixing the document locally. That's why there were missing things - merging through the GitHub interface went wrong because document versions didn't match.

Additionally, this time I've run the text through Grammarly. It found some mistakes, which I corrected. I'm sorry that I didn't do it earlier. It would save you some hassle. I also checked if every link works - and it does.

jglick commented on docs/_tutorials/read-only-json-api.md:
```diff
@@ -0,0 +1,357 @@
+
+By the end of the tutorial, you'll expose the following API endpoints:
+* `/` --- an index of all posts
+* `/{url}` --- details of a post
+* `/categories` --- an index of categories
+* `/categories/{category}` --- an index of posts under a category
```
```suggestion
* `/{url}/` --- details of a post
* `/categories/` --- an index of categories
* `/categories/{category}/` --- an index of posts under a category
```

since [AFAICT](https://stackoverflow.com/a/68562806/12916) you cannot get rid of the slash:

```console
$ curl -IL https://izdwuut.github.io/jekyll-json-api-tutorial/categories
HTTP/2 301 
server: GitHub.com
content-type: text/html
permissions-policy: interest-cohort=()
strict-transport-security: max-age=31556952
location: https://izdwuut.github.io/jekyll-json-api-tutorial/categories/
access-control-allow-origin: *
expires: Wed, 28 Jul 2021 15:46:05 GMT
cache-control: max-age=600
x-proxy-cache: MISS
x-github-request-id: A044:447C:5DE631:890EB7:61017965
accept-ranges: bytes
date: Wed, 28 Jul 2021 15:36:52 GMT
via: 1.1 varnish
age: 47
x-served-by: cache-fty21356-FTY
x-cache: HIT
x-cache-hits: 1
x-timer: S1627486612.323765,VS0,VE1
vary: Accept-Encoding
x-fastly-request-id: 18eb9201adfa28e4b3ee52de9c53090e57201a12
content-length: 162

HTTP/2 200 
server: GitHub.com
content-type: application/json; charset=utf-8
permissions-policy: interest-cohort=()
strict-transport-security: max-age=31556952
last-modified: Thu, 14 Jan 2021 11:45:10 GMT
access-control-allow-origin: *
etag: "60002ec6-15"
expires: Wed, 28 Jul 2021 15:45:47 GMT
cache-control: max-age=600
x-proxy-cache: MISS
x-github-request-id: 5514:447C:5DE3FF:890B33:61017953
accept-ranges: bytes
date: Wed, 28 Jul 2021 15:36:52 GMT
via: 1.1 varnish
age: 65
x-served-by: cache-fty21356-FTY
x-cache: HIT
x-cache-hits: 1
x-timer: S1627486612.347649,VS0,VE1
vary: Accept-Encoding
x-fastly-request-id: e56fa6c0ce7c97e157f7c7ab45c59c92e7fed259
content-length: 21
```
